db/system_keyspace.cc

```diff
@@ -1543,11 +1543,18 @@ future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace:
 }
 
 future<std::vector<gms::inet_address>> system_keyspace::load_peers() {
-    auto res = co_await execute_cql(format("SELECT peer FROM system.{}", PEERS));
+    auto res = co_await execute_cql(format("SELECT peer, tokens FROM system.{}", PEERS));
     assert(res);
 
     std::vector<gms::inet_address> ret;
     for (auto& row: *res) {
+        if (!row.has("tokens")) {
+            // Ignore rows that don't have tokens. Such rows may
+            // be introduced by code that persists parts of peer
+            // information (such as RAFT_ID) which may potentially
+            // race with deleting a peer (during node removal).
```

Comment:
There could be group0 members which don't yet have TOKENS assigned (on bootstrap), don't we want to load their  mapping?

Reply:
@tgrabiec we're still thinking what to do with this part. I'm inclined to think that we should just avoid writing to peers table in raft.

Reply:
@kostja OTOH I'm secretly trying to make RAFT_ID **the** unique ID of a node eventually (Host ID has the replace problem). My plan won't work without persisting the RAFT_ID of known peers somewhere.

We could leave the code in `storage_service::do_update_system_peers_table`. Even if we don't have `persist_initial_raft_address_map`, eventually (after 2 version upgrades) the feature flag in  `do_update_system_peers_table` condition will be always true so `RAFT_ID` will eventually be persisted.

By the way, perhaps we'll have to replace `system.peers` with another table anyway, where the IP is not the key, but one of the regular columns. `system.peers` was defined with the assumption that the IP is the identifier of a node.

Reply:
Yes, we'll replace the state with a new one. I think now you can pretty much roll a dice to make a decision.

Reply:
> Every function in system_keyspace except load_peers() was checking if at least one regular column exists for a peer row before loading that row; e.g. load_host_ids() only loads a row if host_id column is filled and so on. Only load_peers() loaded all rows unconditionally.

I think those methods are checking for their column, so that get_as() doesn't throw if it's null. I'm not sure how is that relevant here? Here we are reading a mapping between IP and host_id, so as long as host_id is present, we have the mapping.

About the race, can't you make `persist_initial_raft_address_map()` go through the synchronization method which storage_service uses?

I don't see a problem with skipping `persist_initial_raft_address_map()` now, but there could be a problem once we switch gossiper to identify nodes by host ids.

Reply:
One problem with relying on gossiper after restart is that this may introduce a small unavailability window on restart. If we want to get rid of all sleeps and make restarts super fast, we don't want to rely on that.

Reply:
> I think those methods are checking for their column, so that get_as() doesn't throw if it's null. I'm not sure how is that relevant here?

When a peer info is persisted in `storage_service::handle_state_normal`, we never write an empty row, we always write it with some column (`host_id`, `data_center`, etc.). `tokens` is persisted last. So if a peer has `tokens`, it means that the peer was fully persisted.

So the idea for the tokens check was: if a row in `system.peers` has `tokens` column engaged, then it's a real peer. If not, it means that either there was a failure when we were persisting the information, or it has been `DELETE`d (but the `DELETE` raced `persist_initial_raft_address_map`, for example). In any case, it's not a fully saved peer, so we don't load it. `tokens` can act as a flag: "this peer is ready".

I agree, it's ugly.

> About the race, can't you make persist_initial_raft_address_map() go through the synchronization method which storage_service uses?

I'm not sure there's anything you could call a "synchronization method".

There's `storage_service::removenode`, which is called from a REST API endpoint (e.g. by `nodetool`). Near the end of `removenode`, there's a call to `storage_service::excise`, which calls `storage_service::remove_endpoint`, which calls `system_keyspace::remote_endpoint`, which `DELETE`s the peer row. This `DELETE` may be potentially followed with an `INSERT` in `persist_initial_raft_address_map`, which would leave us with a row in `system.peers` which has only the key (the IP) and `raft_id` column, but no other columns (in particular, no `tokens`). I proposed we ignore such rows by checking `has("tokens")`, so even if this occurs, we're not affected.

If it happened and we didn't ignore the row, the upgrade procedure would get stuck because it would try to contact this non-existing peer.

The truth is, this will probably never happen in real life anyway, the part of `removenode` before the `excise` call takes too long, `persist_initial_raft_address_map` would have already been finished long before the `excise`. But it could happen e.g. in tests, if we call `removenode` immediately after enabling Raft and `removenode` is super quick, and introduce some rare flakiness (if we didn't have the `tokens` check or something similar).

> I don't see a problem with skipping persist_initial_raft_address_map() now, but there could be a problem once we switch gossiper to identify nodes by host ids.

We can skip it or we can leave it there. I'm merging the PR as it is (with `persist_initial_raft_address_map` included). We can still decide to modify it later if needed.

Reply:
> One problem with relying on gossiper after restart is that this may introduce a small unavailability window on restart. If we want to get rid of all sleeps and make restarts super fast, we don't want to rely on that.

You mean that Raft group 0 could potentially be unavailable for some time (until the information is regossiped). Note that it doesn't imply we have to sleep, we can still send e.g. schema change requests immediately, but they might take longer to finish (until `append_entries` starts being delivered and so on).

Reply:
Yes, we don't have to do that for correctness, but we should do that for availability. We shouldn't accept user requests (by advertising ourselves as UP) before we can serve them, so that we don't hurt our p99. Until then those requests should go to other coordinators.

Reply:
There's one very good reason why we MUST persist IP<->RAFT_ID mappings: remove_from_group0, which is using during removenode and replace operations.

The removenode and replace operations operate in terms of IP and/or Host ID. To remove a server from group 0, we must translate the IP/HostID to the Raft ID - and in this case, the node is down, and there's no guarantee that its IP is available from gossiper.

For example, consider this scenario: we have 3 nodes A, B, C, we shutdown all of them, then restart A and B. We must be able to remove C from group 0. So we must be able to obtain C's Raft ID.

Perhaps we should do even more: should we be re-gossiping the Raft IDs of *other nodes*? If in the future we allow bootstrapping with some nodes dead, for example - how would the bootstrapped node know the dead node's Raft IDs?

Reply:
Opened issue to remember about this: https://github.com/scylladb/scylladb/issues/12083